File: LeetCode/cpp/340_Longest_Substring_with_At_Most_K_Distinct_Characters.cpp
class Solution {
public:
	int lengthOfLongestSubstringKDistinct(string s, int k) {
		if (s.size() <= k) {
			return s.size();
		}

		unordered_map<char, int> um;
		int answer = -1;

		int left = 0;
		int right = 0;

		while (right < s.size()) {
			um[s[right]] = right;
			right++;

			if (um.size() > k) {
				int min_idx = 50002;
				for (auto m : um) {
					min_idx = min(min_idx, m.second);
				}

				um.erase(s[min_idx]);
				left = min_idx + 1;
			}

			answer = max(answer, (right - left));
		}

		return answer;
	}
};

File: LeetCode/java/340_Longest_Substring_with_At_Most_K_Distinct_Characters.java
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int lengthOfLongestSubstringKDistinct(String s, int k) {
        if (s.length() <= k) {
            return s.length();
        }

        Map<Character, Integer> map = new HashMap<>();
        int answer = -1;

        int left = 0;
        int right = 0;

        while (right < s.length()) {
            map.put(s.charAt(right), right);
            right++;

            if (map.size() > k) {
                int minIdx = Collections.min(map.values());

                map.remove(s.charAt(minIdx));
                left = minIdx + 1;
            }

            answer = Math.max(answer, (right - left));
        }

        return answer;
    }
}